great_expectations/render/renderer_configuration.py

```diff
@@ -180,37 +167,148 @@ def __eq__(self, other: Any) -> bool:
+        if "_raw_kwargs" in values and values["_raw_kwargs"]:
+            v += "  "
+            for evaluation_parameter_count in range(len(values["_raw_kwargs"])):
+                v += f" $eval_param__{evaluation_parameter_count},"
+            v = v[:-1]
+        return v
+
+    @staticmethod
+    def _choose_schema_type_for_value(
+        schema_types: List[RendererSchemaType], value: Any
+    ) -> RendererSchemaType:
+        for schema_type in schema_types:
+            try:
+                renderer_param: Type[
+                    BaseModel
+                ] = RendererConfiguration._get_renderer_param_base_model_type(
+                    name="try_param"
+                )
+                renderer_param(schema={"type": schema_type}, value=value)
+                return schema_type
+            except RendererConfigurationError:
+                pass
+        raise RendererConfigurationError(
+            f"None of the schema_types: {schema_types} match the value: {value}"
+        )
+
     def add_param(
         self,
         name: str,
-        schema_type: Union[RendererSchemaType, str],
+        schema_type: Union[RendererSchemaType, List[RendererSchemaType]],
         value: Optional[Any] = None,
     ) -> None:
         """Adds a param that can be substituted into a template string during rendering.
 
         Attributes:
             name (str): A name for the attribute to be added to this RendererConfiguration instance.
-            schema_type (RendererSchemaType or string): The type of value being substituted. One of:
-                - string
-                - number
-                - boolean
-                - array
+            schema_type (list of, or single, RendererSchemaType or string): The possible types for the value being
+                substituted. If more than one schema_type is passed, inference based on param value will be performed,
+                and the first schema_type to match the value will be selected.
+                    One of:
+                     - array
+                     - boolean
+                     - date
+                     - number
+                     - string
             value (Optional[Any]): The value to be substituted into the template string. If no value is
                 provided, a value lookup will be attempted in RendererConfiguration.kwargs using the
                 provided name.
 
         Returns:
             None
         """
-        renderer_param: Type[BaseModel] = create_model(
-            name,
-            renderer_schema=(
-                Dict[str, Optional[RendererSchemaType]],
-                Field(..., alias="schema"),
-            ),
-            value=(Union[Any, None], ...),
-            __base__=RendererConfiguration._RendererParamBase,
-        )
+        renderer_param: Type[
+            BaseModel
+        ] = RendererConfiguration._get_renderer_param_base_model_type(name=name)
```

Comment:
Should be able to make this more specific if the `_get_renderer_param_base_model_type` return type is updated.